config/v1/types_oauth.go

```diff
@@ -1,37 +1,478 @@
+type LDAPAttributeMapping struct {
+	// ID is the list of attributes whose values should be used as the user ID. Required.
+	// LDAP standard identity attribute is "dn"
+	ID []string `json:"id"`
+	// PreferredUsername is the list of attributes whose values should be used as the preferred username.
+	// LDAP standard login attribute is "uid"
+	// +optional
+	PreferredUsername []string `json:"preferredUsername"`
+	// Name is the list of attributes whose values should be used as the display name. Optional.
+	// If unspecified, no display name is set for the identity
+	// LDAP standard display name attribute is "cn"
+	// +optional
+	Name []string `json:"name"`
+	// Email is the list of attributes whose values should be used as the email address. Optional.
+	// If unspecified, no email is set for the identity
+	// +optional
+	Email []string `json:"email"`
+}
+
+// KeystonePasswordIdentityProvider provides identities for users authenticating using keystone password credentials
+type KeystonePasswordIdentityProvider struct {
+	// RemoteConnectionInfo contains information about how to connect to the keystone server
+	OAuthRemoteConnectionInfo `json:",inline"`
+	// Domain Name is required for keystone v3
+	DomainName string `json:"domainName"`
+	// LegacyLookupByUsername flag indicates that user should be authenticated by username, not keystone ID
+	// DEPRECATED - only use this option for legacy systems to ensure backwards compatibiity
+	// +optional
+	LegacyLookupByUsername bool `json:"useKeystoneIdentity"`
+}
+
+// RequestHeaderIdentityProvider provides identities for users authenticating using request header credentials
+type RequestHeaderIdentityProvider struct {
+	// LoginURL is a URL to redirect unauthenticated /authorize requests to
+	// Unauthenticated requests from OAuth clients which expect interactive logins will be redirected here
+	// ${url} is replaced with the current URL, escaped to be safe in a query parameter
+	//   https://www.example.com/sso-login?then=${url}
+	// ${query} is replaced with the current query string
+	//   https://www.example.com/auth-proxy/oauth/authorize?${query}
+	LoginURL string `json:"loginURL"`
+
+	// ChallengeURL is a URL to redirect unauthenticated /authorize requests to
+	// Unauthenticated requests from OAuth clients which expect WWW-Authenticate challenges will be
+	// redirected here.
+	// ${url} is replaced with the current URL, escaped to be safe in a query parameter
+	//   https://www.example.com/sso-login?then=${url}
+	// ${query} is replaced with the current query string
+	//   https://www.example.com/auth-proxy/oauth/authorize?${query}
+	ChallengeURL string `json:"challengeURL"`
+
+	// ClientCA is a reference to a configmap with the trusted signer certs. If empty, no request
+	// verification is done, and any direct request to the OAuth server can impersonate any identity
+	// from this provider, merely by setting a request header.
+	// +optional
```

Comment:
was this optional before?  I'm surprised, but if it is we have to keep it for compatibility.

Reply:
The docs say optional: https://docs.openshift.com/container-platform/3.11/install_config/configuring_authentication.html#RequestHeaderIdentityProvider